libmamba/include/mamba/util/flat_bool_expr_tree.hpp

```diff
@@ -0,0 +1,738 @@
+
+        void push_variable(const variable_type& var);
+        void push_variable(variable_type&& var);
+        void push_operator(const operator_type& op);
+        void push_operator(operator_type&& op);
+        void push_left_parenthesis();
+        void push_right_parenthesis();
+        void finalize();
+
+        [[nodiscard]] auto tree() const& -> const tree_type&;
+        [[nodiscard]] auto tree() && -> tree_type&&;
+
+    private:
+
+        using postfix_parser_type = PostfixParser<variable_type, operator_type>;
+        struct LeftParenthesis
+        {
+        };
+        using operator_or_parenthesis_type = std::variant<operator_type, LeftParenthesis>;
+        using operator_stack_type = std::vector<operator_or_parenthesis_type>;
+
+        postfix_parser_type m_postfix_parser = {};
+        operator_stack_type m_op_stack = {};
+        std::size_t m_parenthesis_level = 0;
+        bool m_expects_op = false;
+        operator_precedence_type m_op_cmp = {};
+
+        template <typename T>
+        void stack_push(T&& elem);
+        auto stack_pop() -> operator_or_parenthesis_type;
+        auto stack_empty() const -> bool;
+        auto stack_top() const -> const operator_or_parenthesis_type&;
+        auto stack_top_is_parenthesis() const -> bool;
+        auto stack_top_is_op_with_greater_precedence_than(const operator_type&) const -> bool;
+
+        template <typename V>
+        void push_variable_impl(V&& var);
+        template <typename O>
+        void push_operator_impl(O&& op);
+    };
+
+    // TODO C++20 this is std::identity
+    struct identity
+    {
+        template <typename T>
+        constexpr auto operator()(T&& t) const noexcept -> T&&;
+    };
+
+    enum struct BoolOperator
+    {
+        logical_and,
+        logical_or
+    };
+
+    template <typename Variable>
+    class flat_bool_expr_tree
+    {
+    public:
+
+        using self_type = flat_bool_expr_tree<Variable>;
+        using operator_type = BoolOperator;
+        using variable_type = Variable;
+        using tree_type = flat_binary_tree<operator_type, variable_type>;
+        using size_type = typename tree_type::size_type;
+
+        flat_bool_expr_tree() = default;
+        flat_bool_expr_tree(const tree_type& tree);
+        flat_bool_expr_tree(tree_type&& tree);
+
+        [[nodiscard]] auto size() const -> size_type;
+        [[nodiscard]] auto empty() const -> bool;
+
+        void clear();
+        void reserve(size_type size);
+
+        template <typename UnaryFunc = identity>
+        [[nodiscard]] auto evaluate(UnaryFunc&& var_evaluator = {}, bool empty_val = true) const
+            -> bool;
+
+    private:
+
+        using idx_type = typename tree_type::idx_type;
+
+        template <typename UnaryFunc>
+        auto evaluate_impl(UnaryFunc& var_evaluator, idx_type idx) const -> bool;
+
+        tree_type m_tree = {};
+    };
+}
+
+#include <cassert>
+#include <iterator>
+#include <stdexcept>
+#include <type_traits>
```

Comment:
Also, the meaning of the files might have changed because of the code preceding these includes. They really need to be at the top. 

Note for the future: the only way to prevent code preceding an include (inline or from an other include) is to use modules, they completely remove the issue. They just are not ready for production for cross-platform projects.

Reply:
Ok, there are some other files to fix in libmamba then :sweat_smile:

Reply:
I didnt spot them but there are then I agree 😬